src/device-manager/java/WeaveDeviceManager-JNI.cpp

```diff
@@ -3995,6 +3997,46 @@ void Java_nl_Weave_DataManagement_WdmClientImpl_beginRefreshData(JNIEnv *env, jo
     }
 }
 
+void Java_nl_Weave_DataManagement_WdmClientImpl_beginFetchEvents(JNIEnv *env, jobject self, jlong wdmClientPtr, jint timeoutSec)
+{
+    WEAVE_ERROR err = WEAVE_NO_ERROR;
+    WdmClient *wdmClient = (WdmClient *)wdmClientPtr;
+
+    WeaveLogProgress(DeviceManager, "beginFetchEvents() called");
+
+    pthread_mutex_lock(&sStackLock);
+    wdmClient->SetEventFetchingTimeout((uint32_t)timeoutSec);
+    err = wdmClient->RefreshData((void *)"FetchEvents", HandleWdmClientComplete, HandleWdmClientError, NULL, true);
+    pthread_mutex_unlock(&sStackLock);
+
+    if (err != WEAVE_NO_ERROR && err != WDM_JNI_ERROR_EXCEPTION_THROWN)
+    {
+        ThrowError(env, err);
+    }
+}
+
+
+jstring Java_nl_Weave_DataManagement_WdmClientImpl_getEventsNative(JNIEnv *env, jobject self, jlong wdmClientPtr)
+{
+    WEAVE_ERROR err = WEAVE_NO_ERROR;
+    WdmClient *wdmClient = (WdmClient *)wdmClientPtr;
+    BytesData bytesData;
+
+    WeaveLogProgress(DeviceManager, "getEventsNative() called");
+
+    err = wdmClient->GetEvents(&bytesData);
+    SuccessOrExit(err);
+
+exit:
+
+    if (err != WEAVE_NO_ERROR && err != WDM_JNI_ERROR_EXCEPTION_THROWN)
+    {
+        ThrowError(env, err);
+    }
+
+    return env->NewStringUTF(reinterpret_cast<const char *>(bytesData.mpDataBuf));
```

Comment:
This looks not right, You mainly consider to create jstring using bytesData.mpDataBuf and bytesData.mDataLen, otherwises, how do you know when string terminate.
for example, std::string((char *)(bytesData.mpDataBuf), bytesData.mDataLen)

Reply:
According to the code, the bytesData is just used to carry `mEventStrBuffer.c_str()` with type `const char *`, as we are obeying the rule of the usage of c_str() here, I don't think we need to have another copy of the data.